src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs

```diff
@@ -809,41 +807,69 @@ private static LiftingResult UserDefinedBinaryOperatorCanBeLifted(TypeSymbol lef
             // SPEC: error occurs.
 
-            // UNDONE: This is a naive quadratic algorithm; there is a linear algorithm that works. Consider using it.
+            int currentBestIndex;
             var candidates = result.Results;
-            for (int i = 1; i < candidates.Count; ++i)
+            for (currentBestIndex = 0; currentBestIndex < candidates.Count; currentBestIndex++)
+            {
+                if (candidates[currentBestIndex].Kind == OperatorAnalysisResultKind.Applicable)
+                {
+                    break;
+                }
+            }
+
+            if (currentBestIndex == candidates.Count)
+            {
+                return;
+            }
+
+            for (int index = currentBestIndex + 1; index < candidates.Count; index++)
             {
-                if (candidates[i].Kind != OperatorAnalysisResultKind.Applicable)
+                if (candidates[index].Kind == OperatorAnalysisResultKind.Inapplicable)
                 {
                     continue;
                 }
 
-                // Is this applicable operator better than every other applicable method?
-                for (int j = 0; j < i; ++j)
+                var better = BetterOperator(candidates[currentBestIndex].Signature, candidates[index].Signature, left, right, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
                 {
-                    if (candidates[j].Kind == OperatorAnalysisResultKind.Inapplicable)
-                    {
-                        continue;
-                    }
-                    var better = BetterOperator(candidates[i].Signature, candidates[j].Signature, left, right, ref useSiteDiagnostics);
-                    if (better == BetterResult.Left)
-                    {
-                        candidates[j] = candidates[j].Worse();
-                    }
-                    else if (better == BetterResult.Right)
-                    {
-                        candidates[i] = candidates[i].Worse();
-                    }
+                    candidates[index] = candidates[index].Worse();
+                }
+                else if (better == BetterResult.Right)
+                {
+                    candidates[currentBestIndex] = candidates[currentBestIndex].Worse();
+                    currentBestIndex = index;
+                }
+            }
+
+            for (int index = 0; index < candidates.Count; index++)
+            {
+                if (index == currentBestIndex || candidates[index].Kind == OperatorAnalysisResultKind.Inapplicable)
+                {
+                    continue;
+                }
+
+                var better = BetterOperator(candidates[currentBestIndex].Signature, candidates[index].Signature, left, right, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
+                {
+                    Debug.Assert(candidates[index].Kind == OperatorAnalysisResultKind.Worse);
```

Comment:
>Debug.Assert(candidates[index].Kind == OperatorAnalysisResultKind.Worse); [](start = 20, length = 73)

After looking at this line more, I believe this assert can fail after all, and we should keep the original assignment. Here is why I think it can fail. In the previous for, when two candidates are compared as neither, neither is marked as Worse, we simply advance index without changing currentBestIndex. After that, some candidate can be better than the current best, the current best is marked as Worse and currentBestIndex is adjusted. This means that, after that for ends, it is possible to have some candidate, other than the current best, that is not marked as Worse and that hasn't been compared with the current best yet. When we perform comparison between them in this loop, that candidate can compare worse than the current best, and the assert will fail.